### PR TITLE
librbd: Return read request early

### PR DESCRIPTION
**Pull request by lixiaoy1**
Finish read requests if the total length is 0.

Signed-off-by: Li, Xiaoyan <xiaoyan.li@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by lixiaoy1** [comment]
@dillaman Please have a look.

File: src/librbd/io/ImageRequest.cc [hunk 1]
```diff
@@ -342,6 +351,17 @@ int ImageReadRequest<I>::clip_request() {
   return 0;
 }
 
+template <typename I>
+bool ImageReadRequest<I>::finish_request_early() {
+  auto total_bytes = ImageRequest<I>::get_total_length();
```
Comment (dillaman) [inline src/librbd/io/ImageRequest.cc:356]: Nit: 
```suggestion
  auto total_bytes = this->get_total_length();
```

File: src/librbd/io/ImageRequest.cc [hunk 2]
```diff
@@ -420,10 +440,7 @@ bool AbstractImageWriteRequest<I>::finish_request_early() {
       return true;
     }
   }
-  uint64_t total_bytes = 0;
-  for (auto& image_extent : this->m_image_extents) {
-    total_bytes += image_extent.second;
-  }
+  auto total_bytes = ImageRequest<I>::get_total_length();
```
Comment (dillaman) [inline src/librbd/io/ImageRequest.cc:443]: Nit:
```suggestion
  auto total_bytes = this->get_total_length();
```

File: src/librbd/io/ImageRequest.h [hunk 3]
```diff
@@ -83,6 +83,8 @@ class ImageRequest {
     m_trace.event("start");
   }
 
+  uint64_t get_total_length();
```
Comment (dillaman) [inline src/librbd/io/ImageRequest.h:86]: Nit:
```suggestion
  uint64_t get_total_length() const;
```

**Comment by mykaul** [comment]
Just wondering  - a zero length read is a common use case?

File: src/librbd/io/ImageRequest.cc (hunk 1 above)
Comment (lixiaoy1) [reply src/librbd/io/ImageRequest.cc]: DONE

File: src/librbd/io/ImageRequest.cc (hunk 2 above)
Comment (lixiaoy1) [reply src/librbd/io/ImageRequest.cc]: DONE

File: src/librbd/io/ImageRequest.h (hunk 3 above)
Comment (lixiaoy1) [reply src/librbd/io/ImageRequest.h]: DONE

**Comment by lixiaoy1** [comment]
> Just wondering - a zero length read is a common use case?

I am not sure about this point. But if we don't check here, we need to do such checks in other places.

**Comment by lixiaoy1** [comment]
@dillaman thanks for your review. Updated based on your comments.

**Pull request merged**